src/libraries/Microsoft.PowerFx.Connectors/ArgumentMapper.cs

```diff
@@ -0,0 +1,192 @@
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+using Contracts = Microsoft.PowerFx.Core.Utils.Contracts;
+
+namespace Microsoft.PowerFx.Connectors
+{
+    // Map arguments between a Swagger File and a Function signature. 
+    // Deals with: 
+    //  - internals, "connectionId"
+    //  - optional, required, default value 
+    //  - computing signature type
+    //
+    // Given a swagger definition, handles:
+    //  - signature information to create a ServiceFunction
+    //  - reverse mapping to map from function signature back to swagger.
+    internal class ArgumentMapper
+    {
+        // All connectors have an internal parameter named connectionId. 
+        // This is handled specially and value passed by connector. 
+        private const string ConnectionIdParamName = "connectionId";
+
+        public readonly IEnumerable<OpenApiParameter> _openApiParameters;
+
+        public ArgumentMapper(IEnumerable<OpenApiParameter> parameters)
+        {
+            _openApiParameters = parameters;
+
+            var requiredParams = new List<OpenApiParameter>();
+            var optionalParams = new List<OpenApiParameter>();
+
+            foreach (var param in parameters)
+            {
+                var name = param.Name;
+
+                HttpFunctionInvoker.VerifyCanHandle(param.In);
+
+                var paramType = param.Schema.ToFormulaType();
+
+                if (param.TryGetDefaultValue(out var defaultValue))
+                {
+                    _parameterDefaultValues[name] = Tuple.Create(defaultValue, paramType._type);
+                }
+
+                if (param.IsInternal())
+                {
+                    // "Internal" params aren't shown in teh signature. So we need some way of knowing what they are.
+                    // connectorId is a special-cases internal parameter, the channel will stamp it.
+                    // Else it have a default value. 
+                    if (name == ConnectionIdParamName || param.HasDefaultValue())
+                    {
+                        continue;
+                    }
+                    else
+                    {
+                        throw new NotImplementedException($"Unexpected internal param {name}");
+                    }
+                }
+
+                if (param.Required)
+                {
+                    requiredParams.Add(param);
+                }
+                else
+                {
+                    optionalParams.Add(param);
+                }
+
+                var options = param.GetOptions();
+                if (options != null)
+                {
+                    var typedName = new TypedName(paramType._type, new DName(name));
+                    _parameterOptions[typedName] = new List<string>(options);
+                }
+            }
+
+            _optionalParamInfo = optionalParams.ConvertAll(x => Convert(x)).ToArray();
```

Comment:
This is calling List.ConvertAll, List.ToArray 
Those are likely more performant than the generic IEnumerable operators because they know the list size.

Reply:
Tried a piece of code to compare the durations with 10M items to see a differnce and ConvertAll might be a winner but the difference is so small it's really hard to be sure.